core/rust/qdbr/src/allocator.rs

```diff
@@ -21,144 +21,488 @@
+    pub fn allocator(&self) -> QdbAllocator {
+        QdbAllocator {
+            mem_tracking: self.mem_tracking.clone(),
+            tagged_used: self.tagged_used.clone(),
+            memory_tag: 65,
+        }
     }
-}
 
-#[allow(dead_code)] // TODO(amunra): remove once in use
-#[cfg(not(test))]
-pub type QdbAllocator = QdbWatermarkAllocator;
+    pub fn rss_mem_used(&self) -> usize {
+        self.mem_tracking.rss_mem_used.load(Ordering::SeqCst)
+    }
 
-#[allow(dead_code)] // TODO(amunra): remove once in use
-#[cfg(test)]
-pub type QdbAllocator = QdbTestAllocator;
-
-#[no_mangle]
-pub extern "system" fn Java_io_questdb_std_Unsafe_QdbAllocator_create(
-    _env: JNIEnv,
-    _class: JClass,
-    rss_mem_limit: *const AtomicUsize,
-    rss_mem_used: *mut AtomicUsize,
-    mem_used: *mut AtomicUsize,
-    tagged_used: *mut AtomicUsize,
-) -> *mut QdbWatermarkAllocator {
-    Box::into_raw(Box::new(QdbWatermarkAllocator::new(
-        rss_mem_limit,
-        rss_mem_used,
-        mem_used,
-        tagged_used,
-    )))
+    pub fn rss_mem_limit(&self) -> usize {
+        self.mem_tracking.rss_mem_limit.load(Ordering::SeqCst)
+    }
+
+    pub fn set_mem_rss_limit(&self, limit: usize) {
+        self.mem_tracking
+            .rss_mem_limit
+            .store(limit, Ordering::SeqCst);
+    }
+
+    pub fn malloc_count(&self) -> usize {
+        self.mem_tracking.malloc_count.load(Ordering::SeqCst)
+    }
+
+    pub fn realloc_count(&self) -> usize {
+        self.mem_tracking.realloc_count.load(Ordering::SeqCst)
+    }
+
+    pub fn free_count(&self) -> usize {
+        self.mem_tracking.free_count.load(Ordering::SeqCst)
+    }
+
+    pub fn tagged_used(&self) -> usize {
+        self.tagged_used.load(Ordering::SeqCst)
+    }
 }
 
-#[no_mangle]
-pub extern "system" fn Java_io_questdb_std_Unsafe_QdbAllocator_destroy(
-    _env: JNIEnv,
-    _class: JClass,
-    allocator: *mut QdbWatermarkAllocator,
-) {
-    if allocator.is_null() {
-        panic!("allocator pointer is null");
+#[cfg(test)]
+mod tests {
+    use crate::allocator::{take_last_alloc_error, AllocFailure, TestAllocatorState};
+    use std::alloc::Allocator;
+    use std::ptr::NonNull;
+
+    #[test]
+    fn test_mem_tracking_size() {
+        assert_eq!(
+            size_of::<crate::allocator::MemTracking>(),
+            6 * size_of::<u64>()
+        );
+    }
+
+    #[test]
+    fn test_alloc_fail() {
+        let tas = TestAllocatorState::new();
+        let allocator = tas.allocator();
+        let too_large = 1024 * 1024 * 1024 * 1024 * 1024; // 1024 TB
+        let layout = std::alloc::Layout::from_size_align(too_large, 16).unwrap();
+        let result = allocator.allocate(layout);
+        assert!(result.is_err());
+        let last_err = take_last_alloc_error().unwrap();
+        assert!(matches!(last_err, AllocFailure::OutOfMemory { .. }));
     }
 
-    drop(unsafe { Box::from_raw(allocator) })
+    #[test]
+    fn test_alloc_with_limit() {
```

Comment:
It would be good to include multi-threaded tests as well

Reply:
Done. There's now 60 threads performing random allocation/dealloc/grow/shrink operations, all racing each other.
They finally finish doing their thing and I track that they go back to zero and count the right number of malloc/realloc/free.